TERRAFORM_VALIDATOR.md

```diff
@@ -0,0 +1,65 @@
+# terraform-validator configuration
+
+These are instruction to configure [terraform-validator](https://github.com/GoogleCloudPlatform/terraform-validator) in CI/CD for the example foundation.
+
+These instructions consider that you have cloned `terraform-example-foundation` repo locally and had already configured the private repos for `gcp-org`, `gcp-environments`, `gcp-networks` and `gcp-projects`.
+
+## Run terraform-validator with tf-wrapper.sh script manually
+
+The user running terraform-validator must have at least these roles to run terraform-validator:
+
+- **Browser** - `roles/browser`
+- **Security Reviewer** - `roles/iam.securityReviewer`
+
+Installation steps:
+
+1. Install terraform-validator locally
+    1. Follow the instructions on the terraform-validator [user guide](https://github.com/forseti-security/policy-library/blob/master/docs/user_guide.md#how-to-use-terraform-validator) to download the latest version.
+    1. Rename `terraform-validator-linux-amd64` to `terraform-validator`
+    1. Update permissions `chmod +x terraform-validator`
+    1. Move terraform-validator binary to a folder in your PATH, like for example `$HOME/.local/bin`
+1. Outside of your clone of terraform-example-foundation repo clone the policy repo `git clone https://github.com/forseti-security/policy-library.git`
+1. Export the path to the policy repo `export policy_library_path=$(pwd)/policy-library`
+1. Change into one of you private repos like `gcp-org`
+1. Run terraform-validator
+    1. Run `./tf-wrapper.sh init prod`
+    1. Run `./tf-wrapper.sh plan prod`
+    1. Run `./tf-wrapper.sh validate  prod ${policy_library_path}`
+
+You should get a result like this:
+
+```
+*************** TERRAFORM VALIDATE ******************
+      At environment: envs/shared
+      Using policy from: /home/foo/policy-library
+*****************************************************
+ERROR: logging before flag.Parse: I0728 23:16:46.428406    3913 validator.go:177] starting 2 workers
+ERROR: logging before flag.Parse: I0728 23:16:46.428862    3913 validator.go:186] worker 0 starting
+ERROR: logging before flag.Parse: I0728 23:16:46.430272    3913 validator.go:186] worker 1 starting
+No violations found.
+```
+
+## Run terraform-validator in Cloud Build
+
+Configuration steps:
+
+1. The cloud build service account will need at least these roles to run terraform-validator:
+   - **Browser** - `roles/browser`
```

Comment:
Should we add these permissions to the bootstrap code? We can add as something in the example foundation to make sure that validator works.

@morgante - do these permissions like right to you? (they are the product of me and @daniel-cit tinkering until we got things running)

Reply:
Hmm, Validator shouldn't need any *new* permissions beyond what the Cloud Build account already has.

Reply:
does it need permissions in the org, to list/get resources / folders / projects etc? By default, cloudbuild will only be able to see things in its own project and we were running into an error where it was throwing errors because it could not enumerate folders/projects.

Reply:
We should be running Validator using the same service account that Terraform runs as.

Reply:
The problem regarding using the terraform service account is this one (from https://github.com/GoogleCloudPlatform/terraform-validator#Auth)

> terraform-validator commands need to be able to authenticate to Google Cloud APIs. This can be done by generating a credentials.json file 

But we create a folder/organization policy that `Disable service account key creation`

Reply:
Terraform Validator runs using Application Default Credentials. That doesn't require downloading service account keys.

Reply:
Right now, all of the terraform runs using service account impersonation which is what @daniel-cit is alluding too. The cloud build service account, by default only has very limited permissions.

Reply:
Got it. I'm still unclear what security is added through that mechanism though and would opt to broaden the Cloud Build account permissions.

Viewer roles should be sufficient.

Reply:
There are a couple of main benefits:
- Separation of concerns, you need to know about the service account in order to use it. An arbitrary Cloud Build job wont run with those privileges. 
- impersonation can be granted to a super admin group (org admins group in our case), so they can act as the service account for debugging without having to download/distribute a key. (for what is a highly privileged account)

It does introduce some integration challenges, which I think we can weigh up over time against these benefits ^

Reply:
That's fair. I remain skeptical of impersonation actually adding real security benefits, but I see the argument. For now we can stick with it and add the necessary viewer roles to the Cloud Build account.